tools/pre-commit.sh

```diff
@@ -1,28 +1,63 @@
 #!/usr/bin/env bash
+# Copy or link this script into .git/hooks/
+#   $ cp "`pwd`/tools/pre-commit.sh" "`pwd`/.git/hooks/pre-commit"
+# This script runs automatically in the project's root directory (parent of .git/).
 
-python -m black --check ./ > /dev/null 2>&1
-need_format=$?
-set -e
-if [ $need_format -ne 0 ]
-then
-    python -m black ./
-    echo Some Python files were formatted
-    echo You need to do git add and git commit again
-    exit $need_format
+EXIT_STATUS=0
+PROJECT_ROOT=`git rev-parse --show-toplevel`
+
+# Determine command to use in order to collect
+# tracked files. Toggle between `dif` and `lstree`
+# using the `ARMORY_CI_TEST` environmental variable.
+# Example:
+#   $ ARMORY_CI_TEST=1 ./tools/pre-commit.sh
+ARMORY_CI_TEST="${ARMORY_COMMIT_HOOK_CI:-0}"
+
+TRACKED_FILES="git --no-pager diff --name-only"
+if [ "${ARMORY_CI_TEST}" -ne 0 ]; then
+    TRACKED_FILES="git --no-pager ls-files"
 fi
-set +e
-
-python -m tools.format_json --check > /dev/null 2>&1
-need_format=$?
-set -e
-if [ $need_format -ne 0 ]
-then
-    python -m tools.format_json
-    echo Some JSON files were formatted
-    echo You need to do git add and git commit again
-    exit $need_format
+[ -z "$TRACKED_FILES" ] && exit 0
+
+
+pushd $PROJECT_ROOT > /dev/null || exit 1
+    ############
+    # Black
+    echo "⚫ - Executing 'black' formatter..."
+    TARGET_FILES=`${TRACKED_FILES} | grep -E '\.py$' | sed 's/\n/ /g'`
+    [ -z "$TARGET_FILES" ] && exit 0
+    python -m black --check --diff --color $TARGET_FILES
+    if [ $? -ne 0 ]; then
+      python -m black $TARGET_FILES
+      echo "⚫ - some files were formatted."
+      EXIT_STATUS=1
+    fi
+
+    ############
+    # Flake8
+    echo "🎱 - Executing 'flake8' formatter..."
+    python -m flake8 --config=.flake8 ./
+    EXIT_STATUS=$?
+
+    ############
+    # JSON Linting
+    echo "📄 - Executing 'json' formatter..."
+    TARGET_FILES=`${TRACKED_FILES} | sed 's/ /\n/g' | grep -E '.*\.json$'`
+    for TARGET_FILE in ${TARGET_FILES}; do
+        python -mjson.tool --sort-keys --indent=4 ${TARGET_FILE} 2>&1 | diff ${TARGET_FILE} -
+        if [ $? -ne 0 ] ; then
+            JSON_PATCH="`python -mjson.tool --sort-keys --indent=4 ${TARGET_FILE}`"
```

Comment:
Does this set an implicit limit on the size of a json file we can format? (E.g., set to the maximum size of a variable in bash)

Reply:
Not at the moment.

Reply:
Added check